Update Verifier documentation

caithagoras opened the pull request:
```
== NO RELEASE NOTE ==
```

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+    * Rewrite queries before execution to ensure that production data is not modified.
+    * Rewrite ``Select`` queries to ``CreateTableAsSelect``
+       * Column names are determined by running the ``Select`` query with ``LIMIT 0``.
+       * Artificial name are used for unnamed columns.
```
typo: name -> names

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+       * Column names are determined by running the ``Select`` query with ``LIMIT 0``.
+       * Artificial name are used for unnamed columns.
+    * Rewrite ``Insert`` and ``CreateTableAsSelect`` queries to have their table names replaced.
+       * Construct a setup query to create the table necessary for a ``Insert`` query.
```
typo: a ``Insert``  -> an ``Insert``

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+        * Control and Test teardown
+    * Queries are subject to timeouts and retries.
+        * Cluster connection failures and transient Presto failures are retried.
+        * Query retries may conceal reliability issues. All occurred Presto query failures, including the retries ones, are recorded.
```
typo: "the retries ones" -> "the retries"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+        * Cluster connection failures and transient Presto failures are retried.
+        * Query retries may conceal reliability issues. All occurred Presto query failures, including the retries ones, are recorded.
+    * Certain query failures are automatically submitted for re-validation, such as partition dropped or table dropped during query.
+    * See `Failure Resolution`_ for automatic resolving query failures.
```
typo: "for automatic resolving query failures" -> "for auto-resolving of query failures"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+
+Column Checksums
+----------------
+For each column in the control/test query, one or more columns maybe generated in the checksum
```
typo: "one or more columns maybe generated " -> "one or more columns are generated"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+    * 4 columns are generated for verification:
+       * Sum of the cardinality
+       * Checksum of the map
+       * Array checksum of the kep set
```
typo: kep -> key

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+       * Array checksum of the kep set
+       * Array checksum of the value set
+* **Row Columns**
+    * Checksum row fields recursively, as according to the type of the fields.
```
typo: "recursively, as according" -> "recursively according"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+Determinism analysis follows the following steps. If a query is found non-deterministic at any
+point, the analysis will conclude.
+
+* A list of non-deterministic catalogs can be specified. If a query references any table from a
```
nit: Would you mention the configuration property to use?

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+
+* A list of non-deterministic catalogs can be specified. If a query references any table from a
+  non-deterministic catalog, the query is considered non-deterministic.
+* Run the control query again, and the compare the results with the initial control query run.
```
typo: "Run the control query again, and the compare" -> "Run the control query again and compare"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+  non-deterministic catalog, the query is considered non-deterministic.
+* Run the control query again, and the compare the results with the initial control query run.
+* If a query has a ``LIMIT n`` clause but no ``ORDER BY`` clause at the top level:
+   * Run a query to count the number of rows produced by the control query without the ``LIMIT`` clause.
```
Does the verifier run the query without the LIMIT clause or with larger value for the LIMIT? Running without a LIMIT may fail if there are too many rows.

caithagoras replied:
For a query `SELECT ... LIMIT n`, Verifier runs
```
SELECT count(*) FROM (SELECT ... )
```
 to get the row count without the limit clause.

It is expensive, and I'm planning to optimize it. This is what we actually need:
```
SELECT count(*) FROM (SELECT ... LIMIT n+1)
```

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+
+* ``EXCEEDED_GLOBAL_MEMORY_LIMIT``: Resolve if the control query uses more memory than the test query.
+* ``EXCEEDED_TIME_LIMIT``: Resolve unconditionally.
+* ``TOO_MANY_HIVE_PARTITIONS``: Resolve if the test cluster do not have enough worker to make sure the
```
typo: "do not have enough worker" -> "does not have enough workers"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+``whitelist``                               A comma-separated list specifying the names of the queries within the suite
+                                            to verify.
+``blacklist``                               A comma-separated list specifying the names of the queries to be excluded
+                                            from suite. ``blacklist`` is applied after ``whitelist``.
```
typo: "from suite" -> "from the suite"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+``blacklist``                               A comma-separated list specifying the names of the queries to be excluded
+                                            from suite. ``blacklist`` is applied after ``whitelist``.
+``source-query-supplier``                   The name of the source query supplier. Supports ``mysql``.
+``source-query.table-name``                 The name of table that holds verifier queries. Available only when
```
typo: "of table" -> "of the table"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+                                            ``stdout``.
+``human-readable.log-file``                 The output files for human-readable events. If not set, human-readable events
+                                            are emitted to ``stdout``.
+``test-id``                                 A String to be attached with output events.
```
typo: "String" -> "string"; "with" -> "to"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+                                            are emitted to ``stdout``.
+``test-id``                                 A String to be attached with output events.
+``max-concurrency``                         Maximum number of concurrent verifications.
+``suite-repetition``                        Number of time a suite is verified.
```
typo: "Number of time" -> "How many times"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+``test-id``                                 A String to be attached with output events.
+``max-concurrency``                         Maximum number of concurrent verifications.
+``suite-repetition``                        Number of time a suite is verified.
+``query-repetition``                        Number of time a source query is verified.
```
typo: "Number of time" -> "How many times"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+                                            floating point column.
+``absolute-error-margin``                   Floating point averages that are below this threshold are treated as ``0``.
+``run-teardown-on-result-mismatch``         Whether to run teardown query in case of result mismatch.
+``verification-resubmission.limit``         Maximum number of time a source query can be re-submitted for verification.
```
typo: "Maximum number of time" -> "A limit on how many times"

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+=========================================== ===============================================================================
+``control.host``                            Host of the control cluster.
+``control.jdbc-port``                       JDBC port of the control cluster.
+``control.http-host``                       Http port of the control cluster.
```
Http -> HTTP

mbasmanova commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,257 @@
+Name                                        Description
+=========================================== ===============================================================================
+``determinism.run-teardown``                Whether to run teardown queries for tables produced in determinism analysis.
+``determinism.max-analysis-runs``           Maximum number of addition control runs to check for the determinism of the
```
addition -> additional

caithagoras commented:
@mbasmanova Thanks for reviewing.
Will leave the PR open for a day in case I get any additional feedback.

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+===============
+
+Presto Verifier is a tool to run queries and verify correctness. It can be used to test whether a
+new Presto version produce the correct query results, or to test if pairs of Presto queries have
```
nit: produce -> produces

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+new Presto version produce the correct query results, or to test if pairs of Presto queries have
+the same semantics.
+
+During each Presto release, Verifier is ran to ensure that there is no correctness regression.
```
ran -> run

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+Extending Verifier
+------------------
+
+Verifier can be extended for further behavioral changes in additional to configuration properties.
```
additional -> addition

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+Verifier Procedures
+-------------------
+
+The following steps summarize the workflow of Verifier.
```
I think using the imperative mood (read, apply, rewrite) for these descriptions is a bit confusing because it's not clear when you're instructing the user to do something vs. when you're just describing what the verifier does. It would be helpful if when you're describing what the verifier does you just used present tense (e.g. Reads the list of source queries, Applies overrides to the catalog, ...), and only use imperative if you're instructing the user to do something (e.g. set up the mysql database)

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+
+Query Override Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The following configurations controls the behavior of query metadata modification before verification starts.
```
control -> controls

rschlussel commented on presto-docs/src/main/sphinx/admin/verifier.rst:
```diff
@@ -0,0 +1,258 @@
+Query Execution Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following configurations controls the behavior of query execution on the control cluster.
```
controls -> control

caithagoras commented:
@rschlussel Thank you for reviewing! Comments addressed.

Also,
- Added `Failure Resolution Configuration` section.
- Added configuration property reference for ``control.table-prefix``and ``test.table-prefix``.
